### ISSUE TITLE
Metrics

### ISSUE THREAD

**Issue by SeanBurford** [folbricht/routedns#80]
With metrics you could graph RouteDNS performance (eg. in Grafana).

The benefits are:

- Less need for logging.  Instead of running at "-l 5" you can see behaviour using counters.
- Ability to add graphing and monitoring.
- Additional insight into memory allocations and usage.

To demonstrate the idea I've created https://github.com/SeanBurford/routedns/commit/3ce505c3b3e18c3e7eddd3db6e4646fded579830

Looking at the golang expvar module, I was concerned that metrics would be exported to everyone who can access the HTTP server (eg. the DoH port).  This turns out to not be the case, since you specify a http mux you have to explicitely add the expvar handler to a path of your choosing.  If the DoHListener config included an optional "vars_path" item users could choose whether to export variables and which path to serve them on.

The output from https://server/routedns/vars which looks like this:

```{"cmdline": ["routedns","config.toml"],
"memstats": ...
"routedns.client.dns.error": {"post": 31},
"routedns.client.dns.query": 146,
"routedns.client.dns.response": {"NOERROR": 114, "NXDOMAIN": 1},
"routedns.client.local-udp.query": 61,
"routedns.client.local-udp.response": {"NOERROR": 46, "NXDOMAIN": 7},
"routedns.listener.srv-doh.drop": 0,
"routedns.listener.srv-doh.error": {},
"routedns.listener.srv-doh.method": {"GET": 1},
"routedns.listener.srv-doh.query": 1,
"routedns.listener.srv-doh.response": {"NXDOMAIN": 1},
"routedns.listener.srv-doh.session": {"tcp": 1},
"routedns.listener.srv-tcp.drop": 0,
"routedns.listener.srv-tcp.header": {},
"routedns.listener.srv-tcp.query": 1,
"routedns.listener.srv-tcp.response": {"NOERROR": 1},
"routedns.listener.srv-tcp.session": {"tcp": 1},
"routedns.listener.srv-udp.drop": 0,
"routedns.listener.srv-udp.header": {"resolve": 7},
"routedns.listener.srv-udp.query": 173,
"routedns.listener.srv-udp.response": {"NOERROR": 159, "NXDOMAIN": 7, "SERVFAIL": 7},
"routedns.listener.srv-udp.session": {"udp": 173},
"routedns.pipeline.local-udp.error": {"idle_timeout": 9, "querytimeout": 8},
"routedns.pipeline.local-udp.maxqlen": 13,
"routedns.pipeline.local-udp.query": 61,
"routedns.pipeline.local-udp.response": {"NOERROR": 54, "NXDOMAIN": 7},
"routedns.router.failover.failover": 5,
"routedns.router.failover.failure": {"dns": 31, "local-udp": 8},
"routedns.router.failover.route": {"dns": 146, "local-udp": 54},
"routedns.router.route-query.route": {"local-udp": 7, "set-ecs": 168}
}
```

**Comment by folbricht** [comment]
Good idea, how about starting a separate listener for it though, rather than using the DoH one. Perhaps a new section in the config to keep all the operational concerns, and one could configure it there as an option, independent of the DNS listeners.

**Comment by SeanBurford** [comment]
Acknowledged.  I just wanted to drop in and agree with moving this to a dedicated section and listener, even though I have no PR change to back it up right now.

**Comment by folbricht** [comment]
Merged in #84
